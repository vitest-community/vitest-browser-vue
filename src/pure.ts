import type { Locator, LocatorSelectors } from '@vitest/browser/context'
import { ComponentMountingOptions, mount, type VueWrapper } from '@vue/test-utils'
import type { DefineComponent } from 'vue'
import { debug, getElementLocatorSelectors, type PrettyFormatOptions } from 'vitest-browser-utils'

type ComponentProps<T> = T extends new (...angs: any) => {
  $props: infer P;
} ? NonNullable<P> : T extends (props: infer P, ...args: any) => any ? P : {};

const mountedWrappers = new Set<VueWrapper>()

export interface Screen<Props> extends LocatorSelectors {
  container: HTMLElement
  baseElement: HTMLElement
  debug(el?: HTMLElement | HTMLElement[] | Locator | Locator[], maxLength?: number, options?: PrettyFormatOptions): void
  unmount(): void
  html(): string
  emitted<T = unknown>(): Record<string, T[]>;
  emitted<T = unknown[]>(eventName: string): undefined | T[];
  rerender(props: Partial<Props>): void
}

export interface ComponentRenderOptions<C, P extends ComponentProps<C>> extends Omit<ComponentMountingOptions<C, P>, 'attachTo'> {
  container?: HTMLElement
  baseElement?: HTMLElement
}

export function render<T, C = T extends ((...args: any) => any) | (new (...args: any) => any) ? T : T extends {
  props?: infer Props;
} ? DefineComponent<Props extends Readonly<(infer PropNames)[]> | (infer PropNames)[] ? {
  [key in PropNames extends string ? PropNames : string]?: any;
} : Props> : DefineComponent, P extends ComponentProps<C> = ComponentProps<C>>(
  Component: T,
  {
    container: customContainer,
    baseElement: customBaseElement,
    ...mountOptions
  }: ComponentRenderOptions<C, P> = {},
): Screen<P> {
  const div = document.createElement('div')
  const baseElement = customBaseElement || customContainer || document.body
  const container = customContainer || baseElement.appendChild(div)

  const wrapper = mount(Component, {
    ...mountOptions,
    attachTo: container,
  })

  // this removes the additional wrapping div node from VTU:
  // https://github.com/vuejs/vue-test-utils-next/blob/master/src/mount.ts#L309
  unwrapNode((wrapper as any).parentElement)

  mountedWrappers.add(wrapper)

  return {
    container,
    baseElement,
    debug: (el = baseElement, maxLength, options) => debug(el, maxLength, options),
    unmount: () => wrapper.unmount(),
    html: () => wrapper.html(),
    emitted: name => wrapper.emitted(name),
    rerender: props => wrapper.setProps(props as any),
    ...getElementLocatorSelectors(baseElement),
  }
}

export function cleanup(): void {
  mountedWrappers.forEach(wrapper => {
    if (wrapper.element?.parentNode?.parentNode === document.body) {
      document.body.removeChild(wrapper.element.parentNode)
    }

    wrapper.unmount()
    mountedWrappers.delete(wrapper)
  })
}

function unwrapNode(node: Element) {
  node.replaceWith(...node.childNodes)
}
