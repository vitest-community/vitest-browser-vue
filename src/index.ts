import { page } from '@vitest/browser/context'
import { beforeEach } from 'vitest'
import { cleanup, render } from './pure'

export { render, cleanup } from './pure'
export type { ComponentRenderOptions, RenderResult } from './pure'

page.extend({
  render,
  [Symbol.for('vitest:component-cleanup')]: cleanup,
})

beforeEach(() => {
  cleanup()
})

declare module '@vitest/browser/context' {
  interface BrowserPage {
    render: typeof render
  }
}
