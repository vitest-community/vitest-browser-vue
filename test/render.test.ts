import { expect, test } from 'vitest'
import { page } from '@vitest/browser/context'
import { render } from '../src/index'
import HelloWorld from './fixtures/HelloWorld.vue'
import Counter from './fixtures/Counter.vue'

test('renders simple component', async () => {
  const screen = render(HelloWorld)
  await expect.element(page.getByText('Hello World')).toBeVisible()
  expect(screen.container).toMatchSnapshot()
})

test('renders counter', async () => {
  const screen = render(Counter, {
    props: {
      initialCount: 1,
    },
  })

  await expect.element(screen.getByText('Count is 1')).toBeVisible()
  await screen.getByRole('button', { name: 'Increment' }).click()
  await expect.element(screen.getByText('Count is 2')).toBeVisible()
})
