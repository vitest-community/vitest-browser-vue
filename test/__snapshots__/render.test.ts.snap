// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders simple component 1`] = `
<div>
  Hello World
</div>
`;
